File: C++/1389_케빈베이컨의6단계법칙_BFS.cpp
#include <iostream>
#include <cstring>
#include <queue>
using namespace std;

int main() {
	int N, M;
	cin >> N >> M;
	
	int **kb = new int*[N]; // kb[i][j]: i->j까지의 최단 경로 길이
	bool **connected = new bool*[N]; // connected[i][j]: i와 j가 연결되었는지 여부
	for(int i = 0; i < N; i++){
		connected[i] = new bool[N];
		memset(connected[i],false,N);
		kb[i] = new int[N];
		memset(kb[i],0,sizeof(int)*N);
	} // 초기화
	
	for(int i = 0; i < M; i++){
		int p1, p2;
		cin >> p1 >> p2;
		connected[p1-1][p2-1] = true;
		connected[p2-1][p1-1] = true;
	} // 연결되어 있음을 표시(방향이 없기 때문에 양쪽 다 표시)
	
	for(int i = 0; i < N; i++){ // i를 시작으로 나머지까지의 최단 거리를 구하고자 함
		queue<pair<int, int> > q;
		q.push(make_pair(i,0)); // 도착지점 및 해당 지점까지의 거리
		
		bool visit[N] = {false, }; // 방문 체크
		visit[i] = true;
		while(!q.empty()){ // BFS 수행
			int num = q.front().first, step = q.front().second;
			q.pop();
			
			for(int j = 0; j < N; j++){
				if(!visit[j] && connected[num][j]){ // 방문한 적 없고 연결된 점 넣기 
					kb[i][j] = step+1;
					kb[j][i] = step+1; // 최단거리
					q.push(make_pair(j,step+1)); // 다음 탐색을 위해 queue에 넣기
					visit[j] = true; // 방문 체크
				}
			}
		}
	}
	
	int p_num = 0, n_kb = 1000000;
	for(int i = 0; i < N; i++){ //케빈 베이컨의 수가 가장 작은 사람 구하기 위함
		int tmp = 0;
		for(int j = 0; j < N; j++){
			tmp += kb[i][j];
		}
		
		if(n_kb > tmp){ // 케빈 베이컨의 수의 최솟 값 및 케빈 베이컨의 수가 가장 작은 사람 갱신(같을 때는 번호가 작은 사람을 출력해야 하므로 아무 것도 수행하지 않음)
			n_kb = tmp;
			p_num = i;
		}
	}
	
	cout << p_num+1; // index는 0부터 시작하므로 1 증가시켜 출력
	
	return 0;
}

File: Python/1389_케빈베이컨의6단계법칙_BFS.py
import sys
from collections import deque

if __name__ == '__main__':
	N, M = list(map(int,sys.stdin.readline().split()))
	
	connected = [[False]*N for _ in range(N)] # connected[i][j]: i와 j가 연결되었는지 나타냄
	dist = [[0]*N for _ in range(N)] # dist[i][j]: i와 j간의 최단 경로

	for _ in range(M):
		p1, p2 = list(map(int,sys.stdin.readline().split()))
		connected[p1-1][p2-1] = True
		connected[p2-1][p1-1] = True # 방향이 없는 그래프임으로 쌍방 연결 표시
		
	for i in range(N): # 시작점 1~N
		q = deque() # BFS를 위한 queue
		q.append([i,0]) # 도착점과 해당 도착점까지의 최단 경로 길이
		
		visit = [False]*N # 방문 체크
		visit[i] = True
		while q: # BFS 시작
			num, step = q.popleft()
			
			for j in range(N):
				if not visit[j] and connected[num][j]: # 방문한적 없고 연결되어있으면
					visit[j] = True # 방문 표시
					q.append([j,step+1]) # 다음 탐색을 위해 queue에 추가
					dist[i][j] = step+1
					dist[j][i] = step+1 # 최단 경로
	
	num = 0
	min_dist = 10000
	for i in range(N): # 케빈 베이컨의 수가 가장 작은 사람을 출력을 위함
		tmp = sum(dist[i])
		if(tmp < min_dist): # 최소 케빈 베이컨의 수 갱신 및 그 때 해당하는 사람(같을 경우에는 수가 번호가 작아야 하므로 갱신하지 않음)
			min_dist = tmp
			num = i
			
	print(num+1) # 인덱스가 0부터 시작함 -> 1 증가하여 출력
